portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -21,29 +21,51 @@
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.*;
+import com.google.gson.Gson;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
-  ArrayList<String> comments = new ArrayList<String>();
-  
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String json = convertToJson(comments);
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery res = datastore.prepare(query);
+
+    List<Entity> comments = new ArrayList<>();
+    for(Entity entity : res.asIterable()) {
+        String content = (String) entity.getProperty("content");
```

Comment:
So the code here feels a bit strange - you have an Entity object, deconstruct it into its parts, then put it back together again to add to the ArrayList. Why not add it right away?

But stepping back for a second - the code on line 54 turns the Entity object (which you don't know the internals of, it being from a 3rd party library) into JSON instead of turning a format you have control over (such as the Comment class you presumably are using in your other PR). This makes it so that your JavaScript below is forced to examine internal properties of the object which could change if the Entity library decided to change, which is brittle.

In summary, I would advise against serializing the Entity object directly to JSON. You can use the Comment object now, or since the only thing your JS uses is the content property, you can just serialize an ArrayList<String> of the content properties.

Let me know if this is confusing.

Reply:
You're not using the `content` and `timestamp` constants you added in line 37-38. And you end up using the same variable name inside your for-loop.

Perhaps, rename those constants to `contentParameter` and `timestampParameter` to avoid confusion. And then use those parameters when grabbing the content/timestamp:
`String content = (String) entity.getProperty(contentParameter);`